lang/src/loader_account.rs

```diff
@@ -29,23 +28,29 @@ use std::ops::DerefMut;
 pub struct AccountLoader<'info, T: ZeroCopy + Owner> {
     acc_info: AccountInfo<'info>,
     phantom: PhantomData<&'info T>,
+    data: *mut T,
 }
 
 impl<'info, T: ZeroCopy + Owner + fmt::Debug> fmt::Debug for AccountLoader<'info, T> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("AccountLoader")
             .field("acc_info", &self.acc_info)
             .field("phantom", &self.phantom)
+            .field("data", unsafe { &*self.data })
             .finish()
     }
 }
 
 impl<'info, T: ZeroCopy + Owner> AccountLoader<'info, T> {
-    fn new(acc_info: AccountInfo<'info>) -> AccountLoader<'info, T> {
-        Self {
+    fn new(acc_info: AccountInfo<'info>) -> Result<AccountLoader<'info, T>, ProgramError> {
+        let data = bytemuck::from_bytes::<T>(&(acc_info.try_borrow_mut_data())?[8..]) as *const T
+            as *mut T;
```

Comment:
I like this proposal a lot more 👍🏻 

However, there's one problem here: `try_borrow_mut_data` is dereferenced, then immediately converted to a pointer, and `RefMut<T>` is dropped at the end of `new`, after which you use a raw pointer. This will side-step the RefCell borrow checking, allowing you to hold multiple mutable references to the underlying data by building several loaders.

I'd suggest you validate the reference via `from_bytes` here, but then store `Ref`/`RefMut` returned from `try_borrow/try_borrow_mut_data` on the AccountLoader instead of the raw pointer. Then you can use bytemuck (or manually via `mem::transmute`) inside Deref/DerefMut.

Reply:
my first attempts did this but I always ran into problems with the borrow checker which I wasn't able to resolve. I then read a paragraph from the rustonomicon writer calling what I'm trying to do (save a reference to myself in myself) a rust cardinal sin and the way they solved it was a raw pointer (https://rust-unofficial.github.io/too-many-lists/fifth-layout.html). but maybe there's a way to do it without a raw pointer and I misunderstood what they said

but lets assume that i could, I think users have to make sure again they free all references before a CPI call. I'd really like to find a way for that not to be the case. Im not sure how much of a problem being able to have multiple mutable loaders is in practice. Loaders are created by anchor so you only ever really have one. you could construct a 2nd one from the same account but idk why anyone would do that and if they do, maybe it's fine to just tell them to be aware of the drawbacks of doing so